File: javaapp/src/main/java/com/scandit/shelf/javaapp/ui/storeselection/StoreSelectionFragment.java
package com.scandit.shelf.javaapp.ui.storeselection;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.material.textfield.TextInputEditText;
import com.scandit.shelf.catalog.Store;
import com.scandit.shelf.javaapp.R;
import com.scandit.shelf.javaapp.ui.base.NavigationFragment;
import com.scandit.shelf.javaapp.ui.pricecheck.PriceCheckFragment;

import java.util.Collections;

public class StoreSelectionFragment extends NavigationFragment implements SwipeRefreshLayout.OnRefreshListener {

    public static StoreSelectionFragment newInstance() {
        return new StoreSelectionFragment();
    }

    private StoreSelectionViewModel viewModel;
    private SwipeRefreshLayout swipeRefreshLayout;
    private TextInputEditText searchView;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewModel = (new ViewModelProvider(this)).get(StoreSelectionViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState
    ) {
        return inflater.inflate(R.layout.store_selection_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        swipeRefreshLayout = rootView.findViewById(R.id.swipe_layout);
        swipeRefreshLayout.setOnRefreshListener(this);
        setupSearchView();

        viewModel.initStoreSelection();

        setupRecyclerView();
        observeLiveData();

        viewModel.refreshStores();
    }

    private void setupSearchView() {
        searchView = rootView.findViewById(R.id.search_view);
        searchView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // Nothing to do
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                viewModel.onSearchTextChanged(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {
                // Nothing to do
            }
        });
    }

    private void setupRecyclerView() {
        MutableLiveData<Store> selectedStoreLiveData = new MutableLiveData<>(null);
        StoresAdapter storesAdapter = new StoresAdapter(getViewLifecycleOwner(), selectedStoreLiveData);

        RecyclerView storesRecyclerView = rootView.findViewById(R.id.stores);
        storesRecyclerView.setHasFixedSize(false);
        storesRecyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
        storesRecyclerView.setAdapter(storesAdapter);
        storesRecyclerView.addItemDecoration(new DividerItemDecoration(requireContext(), RecyclerView.VERTICAL));

        viewModel.storeListLiveData.observe(getViewLifecycleOwner(), it -> {
            Collections.sort(it, (o1, o2) -> o1.getName().compareTo(o2.getName()));
            storesAdapter.submitList(it);
        });

        selectedStoreLiveData.observe(getViewLifecycleOwner(), selectedStore -> {
            if (selectedStore != null) {
                showSnackbar(
                        "Fetching the products for " + selectedStore.getName() + " (id=" + selectedStore.getId() + ")"
                );
                viewModel.refreshProducts(selectedStore);
                hideKeyboard();
            }
        });
    }

    private void observeLiveData() {
        viewModel.isRefreshingLiveData.observe(getViewLifecycleOwner(), swipeRefreshLayout::setRefreshing);

        viewModel.snackbarMessageLiveData.observe(getViewLifecycleOwner(), this::showSnackbar);

        viewModel.storeLiveData.observe(getViewLifecycleOwner(), store ->
        {
            if (store != null) {
                moveToFragment(PriceCheckFragment.newInstance(store.getName()), true, null);
            }
        });
    }

    @Override
    public void onRefresh() {
        searchView.setText("");
        viewModel.refreshStores();
    }

    private void hideKeyboard() {
        InputMethodManager imm = ContextCompat.getSystemService(requireContext(), InputMethodManager.class);
        if (imm != null) {
            imm.hideSoftInputFromWindow(requireView().getWindowToken(), 0);
        }
    }
}

File: kotlinapp/src/main/kotlin/com/scandit/shelf/kotlinapp/ui/storeselection/StoreSelectionFragment.kt
package com.scandit.shelf.kotlinapp.ui.storeselection

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import androidx.core.content.ContextCompat
import androidx.core.widget.doOnTextChanged
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.textfield.TextInputEditText
import com.scandit.shelf.catalog.Store
import com.scandit.shelf.kotlinapp.R
import com.scandit.shelf.kotlinapp.ui.base.NavigationFragment
import com.scandit.shelf.kotlinapp.ui.pricecheck.PriceCheckFragment
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.launch

class StoreSelectionFragment : NavigationFragment(), SwipeRefreshLayout.OnRefreshListener {

    private lateinit var viewModel: StoreSelectionViewModel
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private lateinit var searchView: TextInputEditText

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProvider(this)[StoreSelectionViewModel::class.java]
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View = inflater.inflate(R.layout.store_selection_fragment, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        swipeRefreshLayout = rootView.findViewById(R.id.swipe_layout)
        swipeRefreshLayout.setOnRefreshListener(this)
        setupSearchView()

        viewModel.initStoreSelection()

        setupRecyclerView()
        collectFlows()

        viewModel.refreshStores()
    }

    private fun setupSearchView() {
        searchView = rootView.findViewById<TextInputEditText>(R.id.search_view)
            .apply {
                doOnTextChanged { text, _, _, _ -> viewModel.onSearchTextChanged(text.toString()) }
            }
    }

    private fun setupRecyclerView() {
        val selectedStoreChannel = Channel<Store>()
        val storesAdapter = StoresAdapter(viewLifecycleOwner, selectedStoreChannel)

        rootView.findViewById<RecyclerView>(R.id.stores).apply {
            setHasFixedSize(false)
            layoutManager = LinearLayoutManager(requireContext())
            adapter = storesAdapter
            addItemDecoration(DividerItemDecoration(requireContext(), RecyclerView.VERTICAL))
        }

        viewLifecycleOwner.lifecycleScope.launchWhenResumed {
            viewModel.storeListFlow.collectLatest {
                storesAdapter.submitList(it.sortedBy { store -> store.name })
            }
        }

        viewLifecycleOwner.lifecycleScope.launch {
            selectedStoreChannel.receiveAsFlow()
                .collect {
                    showSnackbar("Fetching the products for ${it.name} (id=${it.id})")
                    viewModel.refreshProducts(it)
                    hideKeyboard()
                }
        }
    }

    private fun collectFlows() {
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.isRefreshingFlow.collectLatest(swipeRefreshLayout::setRefreshing)
        }

        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.snackbarMessageFlow.filterNotNull().collectLatest(::showSnackbar)
        }

        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.storeFlow.filterNotNull().collectLatest { store ->
                moveToFragment(
                    PriceCheckFragment.newInstance(store.name),
                    true,
                    null
                )
            }
        }
    }

    override fun onRefresh() {
        searchView.setText("")
        viewModel.refreshStores()
    }

    companion object {
        fun newInstance() = StoreSelectionFragment()
    }
}

private fun Fragment.hideKeyboard() {
    ContextCompat.getSystemService(requireContext(), InputMethodManager::class.java)
        ?.hideSoftInputFromWindow(requireView().windowToken, 0)
}
